lls/cache.c

```diff
@@ -395,8 +385,96 @@ xmit_icmp_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
 	icmph->icmp_code = ICMP_ECHO_REPLY_CODE;
 	icmph->icmp_cksum = 0;
 	icmph->icmp_cksum = icmp_cksum(icmph, sizeof(*icmph));
+}
+
+static void
+process_icmp_ping_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
+	struct lls_config *lls_conf, struct gatekeeper_if *iface)
+{
+	uint16_t tx_queue;
+	struct token_bucket_ratelimit_state *rs;
+	unsigned int i;
+	unsigned int num_granted_pkts;
+
+	if (iface == &lls_conf->net->front) {
+		tx_queue = lls_conf->tx_queue_front;
+		rs = &lls_conf->front_icmp_rs;
+	} else {
+		tx_queue = lls_conf->tx_queue_back;
+		rs = &lls_conf->back_icmp_rs;
+	}
+
+	num_granted_pkts = tb_ratelimit_allow_n(num_pkts, rs);
+
+	for (i = 0; i < num_granted_pkts; i++)
+		xmit_icmp_ping_reply(iface, pkts[i]);
+	send_pkts(iface->id, tx_queue, num_granted_pkts, pkts);
+
+	/* XXX #435: Adopt rte_pktmbuf_free_bulk() when DPDK is updated. */
+	for (i = num_granted_pkts; i < num_pkts; i++)
+		rte_pktmbuf_free(pkts[i]);
 
-	return 0;
+}
+
+static void
+process_icmp_pkts(struct lls_config *lls_conf, struct lls_icmp_req *icmp)
+{
+	struct rte_mbuf *ping_pkts[icmp->num_pkts];
+	unsigned int num_ping_pkts = 0;
+	int i;
+
+	for (i = 0; i < icmp->num_pkts; i++) {
+		struct rte_mbuf *pkt = icmp->pkts[i];
+		struct rte_ether_hdr *eth_hdr =
+			rte_pktmbuf_mtod(pkt, struct rte_ether_hdr *);
+		struct rte_ipv4_hdr *ip4hdr;
+		struct rte_icmp_hdr *icmphdr;
+		size_t l2_len = pkt_in_l2_hdr_len(pkt);
+
+		pkt_in_skip_l2(pkt, eth_hdr, (void **)&ip4hdr);
+		if (pkt->data_len < (ICMP_PKT_MIN_LEN(l2_len) +
+				ipv4_hdr_len(ip4hdr) - sizeof(*ip4hdr))) {
+			rte_pktmbuf_free(pkt);
+			continue;
+		}
+
+		/*
+		 * We must check whether the packet is fragmented here because
+		 * although match_icmp() checks for it, the ACL rule does not.
+		 */
+		if (rte_ipv4_frag_pkt_is_fragmented(ip4hdr)) {
+			LLS_LOG(WARNING,
+				"Received fragmented ICMP packets destined to this server at %s\n",
+				__func__);
+			rte_pktmbuf_free(pkt);
+			continue;
+		}
+
+		/*
+		 * We don't need to make sure the next header is ICMP
+		 * because both match_icmp() and the ACL rule already check.
+		 */
+
+		icmphdr = (struct rte_icmp_hdr *)ipv4_skip_exthdr(ip4hdr);
+
+		if (icmphdr->icmp_type == ICMP_ECHO_REQUEST_TYPE &&
+				icmphdr->icmp_code == ICMP_ECHO_REQUEST_CODE) {
+			ping_pkts[num_ping_pkts++] = pkt;
+		} else if (icmphdr->icmp_type == ICMP_DEST_UNREACHABLE_TYPE &&
+				icmphdr->icmp_code == ICMP_FRAG_REQ_DF_CODE) {
+			LLS_LOG(ERR, "Received \"Fragmentation required, and DF flag set\" ICMP packet on %s interface; check MTU along path\n",
+				icmp->iface->name);
+			rte_pktmbuf_free(pkt);
+		} else {
+			LLS_LOG(INFO, "Received ICMP packet with type %hhu and code %hhu on %s interface\n",
+				icmphdr->icmp_type, icmphdr->icmp_code, icmp->iface->name);
+			rte_pktmbuf_free(pkt);
+		}
```

Comment:
Reporting the source address in `ip4hdr` in both log messages would be valuable in production. Add `continue;` after line `ping_pkts[num_ping_pkts++] = pkt;`, obtain a string for the source address, and add it to both messages. After this change, `rte_pktmbuf_free(pkt);` can be placed outside of the new `if` statement.

This comment applies to the IPv6 patch as well.